File: LeetCode 热题 HOT 100/二叉树/199. 二叉树的右视图.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var rightSideView = function (root) {
  ans = []
  if (root === null) {
    return ans
  }
  res = []
  function getVal(node) {
    if (node.left) {
      points.push(node.left)
    }

    if (node.right) {
      points.push(node.right)
    }
  }

  points = []
  points.push(root)
  while (points.length !== 0) {
    size = points.length
    result = []
    for (let index = 0; index < size; index++) {
      temp = points.shift()
      result.push(temp.val)
      getVal(temp)
    }
    res.push(result)
  }

  for (let i in res){
    ans.push(res[i][res[i].length - 1])
  }
  return ans
};

File: LeetCode 热题 HOT 100/二叉树/199. 二叉树的右视图.py
# 给定一个二叉树的 根节点 root，想象自己站在它的右侧，按照从顶部到底部的顺序，返回从右侧所能看到的节点值。

# 示例 1:
# 输入: [1, 2, 3, null, 5, null, 4]
# 输出: [1, 3, 4]

# 示例 2:
# 输入: [1, null, 3]
# 输出: [1, 3]

# 示例 3:
# 输入: []
# 输出: []


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

from collections import deque


class Solution:
    def rightSideView(root):
        ans = []
        if not root:
            return ans
        res = []

        def getVal(node):
            if node.left:
                points.append(node.left)
            if node.right:
                points.append(node.right)

        points = deque([root])
        while points:
            result = []
            size = len(points)
            for _ in range(size):
                temp = points.popleft()
                result.append(temp.val)
                getVal(temp)
            res.append(result)

        for i in range(len(res)):
            ans.append(res[i][-1])
        return ans
